[Balasubramaniam Praveen] iP

pr4aveen opened the pull request.

Commit:
Add Gradle support

pr4aveen committed:
Add Level-1 Requirements

Add features for Level-1 of Duke as specified in the project
requirements.

Let's
 * Make the program start by greeting the user
 * Make the program echo commands entered by the user
 * Make the program exit when the user types 'bye'

See below for more details:
https://nus-cs2103-ay2021s1.github.io/website/admin/ip-w2.html

pr4aveen committed:
Read Commands: Use switch-case statements instead of if-else

Program uses if-else statements to read commands.

Adding more commands in the future will make the code less readable.

Let's use switch-case statements instead to make future code more
readable.

pr4aveen committed:
Add Level-2 Requirements

Program echos all user input except 'bye'.

Program should store user input and echo them when the command 'list' is
 used.

Let's
 * Make the program store all user input that is not 'list' or 'bye'
 * Make the program list all previous entries when 'list' is used

See below for more details:
https://nus-cs2103-ay2021s1.github.io/website/admin/ip-w2.html

pr4aveen committed:
Create task object

Program stores all tasks as Strings.

Tasks should be represented as a separate object so that more
information can be stored within each task.

Let's
 * Create a Task object to represent individual tasks

pr4aveen committed:
Replace string based tasks with task object

Program stores all tasks as strings.

Tasks should be represented with the task object so that more
information can be stored within each task.

Let's
 * Update Duke.java to use the Task object to represent individual tasks

pr4aveen committed:
Add Level-3 Requirements

Program does not allow tasks to be marked as done.

Program should store new tasks as 'not done' and provide users with the
ability to mark tasks as done with the 'done id' command, where id is
the task number.

Let's
 * Make the program mark tasks as done with the 'done' command

See below for more details:
https://nus-cs2103-ay2021s1.github.io/website/admin/ip-w2.html

pr4aveen committed:
Create Deadline, Event and Todo classes

Program only has one type of task.

Program should be able to support deadline, event and todo tasks in
order to fulfil Level-4 requirements.

Let's create classes for each type of task and convert Task.java to an
abstract class.

pr4aveen committed:
Add Level-4 Requirements

Program only has one type of task.

Program should be able to create different types of tasks that take in
different parameters in order to fulfil Level-4 requirements. Deadline,
Event and ToDo tasks should be supported.

Let's
 * Make the program create a Deadline task with the 'deadline dueTime'
 command, where dueTime is the time that the task is due
 * Make the program create an Event task with the 'event startTime'
 command, where startTime is the time that the event will start
 * Make the program create a ToDo task with the 'todo' command

See below for more details:
https://nus-cs2103-ay2021s1.github.io/website/admin/ip-w2.html

pr4aveen committed:
Fix 'list' command on empty list

Fix a bug where the list command crashes the program when the list is
empty. Empty lists will now show the message "You have no tasks!"

pr4aveen committed:
Add automated testing

Program currently needs all commands to be tested manually after a new
command is added.

Let's add automated text ui testing using I/O redirection.

pr4aveen committed:
Refactor task classes to use static factory methods

Program uses public constructors to create new tasks.

Program should be able to perform a series of checks before the creation
 of a new task object. These checks should be self-contained in order to
  improve code readability.

Let's
* Modify the task classes to use static factory methods instead of
public constructors.
* Change access modifier of the Task constructor to protected
* Change access modifiers of ToDo, Event and Deadline to private

pr4aveen committed:
Add Level-5 Requirements

Program has minimal error handling.

Program should be able to perform checks before creating different types
 of tasks and throw an exception if there are any errors.

Let's
* Create a custom exception class called DukeException
* Wrap main loop of the program in a try-catch block
* Throw exceptions from static factory methods and other helper methods
that may be called in the switch-case block

See below for more details:
https://nus-cs2103-ay2021s1.github.io/website/admin/ip-w2.html

pr4aveen committed:
Add Level-6 Requirements

Program does not allow tasks to be deleted.

Program should provide users with the ability to delete tasks with the
'delete id' command, where id is the task number.

Let's
* Make the program delete tasks with the 'delete' command
* Update the necessary tests

pr4aveen committed:
Update diff command in runtest.sh

pr4aveen committed:
Complete A-Enums

Main switch-case statement uses case-insensitive strings to represent
each command.

Let's represent different commands with enums to make commands case
insensitive and make code cleaner.

pr4aveen committed:
Implement Storage class

Program is supposed to be a task manager but does not save information
after each session.

Let's
 * Implement a storage class that creates a file to store information
 if it does not already exist.
 * Allow the storage class to append new information to the file
 whenever required.

pr4aveen committed:
Allow tasks to be read from storage

Program is supposed to be a task manager but does not save information
after each session.

Let's allow the program to create ToDo, Event and Deadline tasks from
the data in local storage

pr4aveen committed:
Allow tasks to be saved to storage

Program is supposed to be a task manager but does not save information
after each session.

Let's allow the program to save ToDo, Event and Deadline tasks to local
storage.

pr4aveen committed:
Add Level 7 requirements

Program is supposed to be a task manager but local storage does not
support marking tasks as done and deleting tasks.

Let's
 * Allow the program to mark ToDo, Event and Deadline tasks as done in
 local storage
 * Allow the program to delete ToDo, Event and Deadline tasks from local
  storage

pr4aveen committed:
Add ability to parse date and time

Program saves event and deadline task's date and time as a string. This
prevents the program from performing more advanced operations on the
tasks such as searching by date.

Let's allow the program to parse and store the date and time of events
and deadlines using the yyyy-MM-dd[ HH[:]mm] format, where the square
brackets represent optional parameters.

pr4aveen committed:
Add Level 8 Requirements

Program parses and stores dates for Event and Deadline tasks.

There is however no way to filter tasks based on the date they are due.

Let's add a date command which takes in a date in the form yyyy-MM-dd as
 its argument and returns a list of tasks due on that day.

pr4aveen committed:
Merge branch 'branch-Level-7'

pr4aveen committed:
Merge branch 'branch-Level-8'

pr4aveen committed:
Update gitignore to exclude data folder

pr4aveen committed:
Delete data directory

pr4aveen committed:
Extract TaskList to TaskList.java

Program currently has the task list and it's operations in Duke.java.

This clutters up Duke.java and may violate certain OOP principles.

Let's extract the TaskList and all related operations to TaskList.java
and change the code in Duke.java accordingly.

pr4aveen committed:
Extract UI Operations to UI.java

All UI related operations are performed in Duke.java.

This clutters up Duke.java and may violate certain OOP principles.

Let's extract all UI related operations to UI.java and change the code
in Duke.java accordingly.

pr4aveen committed:
Add A-MoreOOP

Codebase does not adhere to OOP standards.

The codebase should be modified to make it easier to manage and to
reduce the likelihood of errors in the future.

Let's:
 * Create UI, Parser, Storage classes to handle different operations
 * Create an abstract Command class with a subclass for each command

pr4aveen committed:
Add A-Packages

Files in codebase are not organized.

The class files should be put into different packages to make the
codebase more organized.

Let's put all files into a duke package and create a duke.command
package for command related classes and duke.task for task related
classes.

pr4aveen committed:
Update exception messages

Certain exception messages were altered during refactoring.

Let's revert these modified exception messages back to their original
form.

pr4aveen committed:
Add JUnit Tests

Codebase currently does not have any tests to test the behavior of the
code.

Let's add JUnit tests to test the exceptions thrown during the creation
of new ToDo, Event and Deadline tasks

pr4aveen committed:
Add A-Jar

Project does not have any releases.

Let's set up the project to export releases as JAR files.

pr4aveen committed:
Update gitignore to exclude test data files

pr4aveen committed:
Add A-JavaDoc

Codebase does not contain any form of documentation.

This makes it difficult for people unfamiliar with the codebase to
understand what different methods do.

Let's add some JavaDocs to make document the actions of different public
 methods.

pr4aveen committed:
Add A-CheckStyle

Codebase does not adhere to all coding style guidelines

Let's modify the code such that it adheres to these guidelines.

pr4aveen committed:
Add Level-9

Program currently does not support filtering tasks by a string.

This makes it harder to find tasks once the task list gets long.

Let's add a find command that allows users to search for a task with a
keyword.

pr4aveen committed:
Merge branch 'branch-A-Javadoc'

pr4aveen committed:
Merge branch 'branch-A-CodingStandard'

pr4aveen committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/TaskList.java

pr4aveen committed:
Fix merge conflict bugs

Merge of branch-A-JavaDoc, branch-A-CheckStyle and branch-A-Level-9
introduced a few issues with the code.

Let's
 * Fix a typo in DateCommand.execute JavaDoc
 * Fix an issue with access modifier in Task.java
 * Fix a missing import statement in Parser.java

pr4aveen committed:
Update JavaDoc to include header comments

Classes do not have header comments.

This makes it difficult for users to find out the purpose of each class
file.

Let's update the JavaDocs such that they include header comments.

pr4aveen committed:
Merge branch 'origin/add-gradle-support'

pr4aveen committed:
Add A-Gradle

Project requires users to manually execute JUnit tests, cleanups,
and builds.

This is time consuming.

Let's use gradle to automate the tasks above.

pr4aveen committed:
Merge branch 'branch-A-Gradle'

pr4aveen committed:
Add checkstyle files

Program does not have checkstyle files that are configured to follow the
coding standard.

This is required for us to incorporate checkstyle into gradle.

Let's add the required checkstyle files to the project root directory.

pr4aveen committed:
Update checkstyle version to v8.32

Checkstyle version in build.gradle was 8.23.

This is incompatible with the version of the checkstyle file provided.

Let's update the checkstyle version to 8.32.

pr4aveen committed:
Edit code to comply with checkstyle v8.32

Checkstyle highlights multiple violations of coding standards.

This is not good coding practice and the code will look inconsistent.

Let's fix these violations to make the code look more consistent.

pr4aveen committed:
Merge branch 'branch-A-CheckStyle'

pr4aveen committed:
Update Duke.java to use public constructor

Duke.java currently uses a static factory method.

This might not be compatible with the initial set up of Java FX.

Let's replace the static factory method in Duke.java with a public
constructor and make the necessary changes in preparation for Level-10.

pr4aveen committed:
Add Java FX dependencies to build.gradle

pr4aveen committed:
Add GUI using Java FX

Program is currently CLI based.

This makes it difficult to be used as a personal chat bot application.

Let's create a GUI for the program using Java FX.

pr4aveen committed:
Add GUI auto close on 'bye' command

GUI currently does not close when the bye command is keyed in.

This defeats the purpose of the bye command.

Let's make the bye command terminate the app after a short delay.

pr4aveen committed:
Merge branch 'branch-Level-10'

ypinhsuan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return True if data successfully populated. False otherwise.
+     */
+    public boolean loadFromStorage() {
```
Maybe you can make the boolean method sound more like booleans?

ypinhsuan commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+package duke;
+
```
Should there be javadoc comments here?

ypinhsuan commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+        getChildren().setAll(tmp);
+        setAlignment(Pos.TOP_LEFT);
+    }
+
```
Perhaps it will be nice to have javadoc comments here since this is a public method?
I noticed the same issue in several other places too

ypinhsuan commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,12 @@
+public class DukeException extends Exception {
+
+    /**
+     * Throws a new DukeException
```
Should there be a full stop?

ypinhsuan commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,62 @@
+            return new DeadlineCommand(argument);
+        case FIND:
+            return new FindCommand(argument);
+        case INVALID: default:
```
Maybe the default case should be on the next line?

ypinhsuan commented on src/main/java/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,43 @@
+import duke.UI;
+import duke.task.TaskList;
+import duke.task.ToDo;
+
```
Perhaps the header comments for the class is at the wrong place?

ypinhsuan commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,109 @@
+
+public class Deadline extends Task {
+
+    /**
+     * Encapsulates data and methods specific to Deadline tasks.
+     */
```
Maybe the javadoc comment is at the wrong place?

ypinhsuan reviewed (commented):
Looks good! I like the part where you used CompletableFuture. Overall, your code is clean and easy to read.

JoeyToh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    /**
+     * Creates a new instance of the Duke chat bot with a custom filepath.
+     * @param filePath File path where Duke should read from and save to.
```
Very minor style violation here, but I think there should be an empty line between the description and parameter.

JoeyToh commented on src/main/java/duke/task/TaskList.java:
```diff
@@ -0,0 +1,210 @@
+
+                if (entry.length == 3) {
+                    String taskType = entry[0];
+                    if (!entry[1].toUpperCase().equals("TRUE") && !entry[1].toUpperCase().equals("FALSE")) {
+                        throw new DukeException("One or more task statuses are not stored correctly");
```
Line 171 and 172 seem a bit long so perhaps you could do some line wrapping?

JoeyToh reviewed (commented):
Overall, I think your code is very clean and OOP has been done very well. 💯 
The logic in your code also flows very nicely.
There are only very minor coding standard violations in your Javadocs, most of which have already been commented on by the other reviewer. Nonetheless, the content of your Javadocs is still concise and easy to understand. :)

pr4aveen committed:
Add A-Assertions

Code does not have any assertions.

This is not a good defensive coding practice as it makes it harder to
trace bugs

Let's add assertions to all classes and methods in duke.Duke that
require them.

pr4aveen committed:
Add A-CodeQuality

The loadDataFromStorage() method in task list violates the code quality
guidelines.

This makes the code less understandable.

Let's edit TaskList.java to make it comply with the code quality
guidelines.

pr4aveen committed:
Add A-CI

Repository does not support Continuous Integration (CI).

Let's add a simple CI script to the repository.

pr4aveen committed:
Move gradle.yml to .github/workflows

pr4aveen committed:
Remove trailing whitespace

IDE included whitespace in all empty lines.

This is a violation of the coding standard.

Let's remove all trailing whitespace so that the code will pass
checkstyle.

pr4aveen committed:
Improve find command

The Find command can only filter by keywords found in the task string.
It cannot filter by type or status.

This makes it harder for users to filter by type or status as they will
have to search for keywords such as [T].

Let's improve the find command to make it support more keywords to
filter the search by.

pr4aveen committed:
Update build.gradle

build.gradle file does not enable assertions and has the wrong
mainClassName.

Let's edit the mainClassName and enable assertions in build.gradle

pr4aveen committed:
Add more assertions

pr4aveen committed:
Merge branch 'branch-A-Assertions'

pr4aveen committed:
Refactor repeated code

DateTime parsing code is repeated across the Event and Deadline
classes.

This is violation of the DRY principle.

Let's shift the DateTime parsing to a new class.

pr4aveen committed:
Merge branch 'branch-A-CodeQuality'

pr4aveen committed:
Add C-Help

New users do not have any means of knowing what commands the
application supports.

This makes the app less user friendly.

Let's add a Help command that displays the available commands and
their syntax.

pr4aveen committed:
Merge pull request #1 from pr4aveen/branch-C-Help

Add C-Help

pr4aveen committed:
Remove wildcard imports

Parser class uses wildcard imports.

This is a violation of checkstyle.

Let's update Parser.java to use single class imports instead of wildcard
imports.

pr4aveen committed:
Update GUI

Text is currently displayed without a background.

This is not aesthetically pleasing.

Let's display text in chat bubbles.

pr4aveen committed:
Update user guide

Application does not have a user guide.

This makes it difficult for users to learn more about the application.

Let's add a user guide for the application.

pr4aveen committed:
Set theme jekyll-theme-cayman

pr4aveen committed:
Update help command

Help command displays a list of all commands.

This means that every time the user guide is updated, the information
displayed by the help command needs to be updated as well. The
information displayed on the app is also limited as compared to the
user guide.

Let's redirect the user to the user guide when the help command is used.

pr4aveen committed:
Merge remote-tracking branch 'origin/master'

pr4aveen committed:
Update GUI

Current GUI lacks customization.

This makes the app harder to differentiate from others.

Let's add a background and some more colours to the GUI.

pr4aveen committed:
Rename Duke to Alfred

Project is currently named Duke.

This does not suit the theme of the GUI.

Let's rename Duke to Alfred to better suit the theme.

pr4aveen committed:
Update acknowledgements

pr4aveen committed:
Code cleanup

pr4aveen committed:
Update README.md

pr4aveen committed:
Fix chat bubble cropping bug

Chat bubble crops when the list is over a certain length.

This hides the full message content.

Let's ensure that the chat bubble shows the full message.

pr4aveen committed:
Update GUI

pr4aveen committed:
Fix typos and formatting in README.md
